src/Build/Evaluation/Expander.cs

```diff
@@ -3694,6 +3703,48 @@ private bool TryExecuteWellKnownFunctionWithPropertiesParam(IPropertyProvider<T>
                 return false;
             }
 
+            private object GetMethodResult(object objectInstance, IEnumerable<MethodInfo> methods, object[] args, int index)
+            {
+                for (int i = index; i < args.Length; i++)
+                {
+                    if (args[i].Equals("out _"))
+                    {
+                        object toReturn = null;
+                        foreach (MethodInfo method in methods)
+                        {
+                            Type t = method.GetParameters()[i].ParameterType;
+                            args[i] = t.IsValueType ? Activator.CreateInstance(t) : null;
+                            object currentReturnValue = GetMethodResult(objectInstance, methods, args, i + 1);
+                            if (currentReturnValue is not null)
+                            {
+                                if (toReturn is null)
```

Comment:
This is a bit nitpicky - but I believe we should distinguish between the return valua itself and the fact if it was set or not - as `null` can still be a valid return value and we can have a case of 2 matching functions one of which returns `null` and other non-null value.

**tl;dr;**: How about introducing `bool isReturnSet;` ?

Reply:
I had to think about this a bit because although that's often true, MSBuild code doesn't have a 'null' type, and this goes almost directly to MSBuild. But what you're really saying is that we might discard a null return value incorrectly if the method really did just return null but was correctly invoked, and I think we could fail to distinguish between those.

I'm thinking about what it will ultimately look like when it ultimately gets back to the user. Perhaps a simpler fix to this would be changing line 3739 to end with `?? "null";` instead? How does that sound?

Reply:
My main concern is - you are using `toReturn` here as a way to find out if possibly more distinct overloads of a method - which is unsupported and failure case. But the additional overload would slip through here if it'd return `null`. So in such case there would be 2 similar methods, returning distinct returns, but code would silently ignore one of them (the one returning the null) and accept the other as if it'd be the only matching one.

Ping me if I'm making you even more confused ;-)

Reply:
Works as well